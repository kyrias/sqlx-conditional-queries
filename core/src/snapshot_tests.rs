use proc_macro2::TokenStream;

use crate::DatabaseType;

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

fn prettyprint(ts: TokenStream) -> String {
    // `prettyplease` operates on full files so we have to wrap the generated output in a dummy
    // function.
    let ts = quote::quote! {
        fn dummy() {
            #ts
        }
    };
    let file = syn::parse2(ts).expect("failed to parse wrapped output as a file");
    prettyplease::unparse(&file)
}

#[rstest::rstest]
#[case::postgres(DatabaseType::PostgreSql, true)]
#[case::postgres_unchecked(DatabaseType::PostgreSql, false)]
#[case::mysql(DatabaseType::MySql, true)]
#[case::mysql_unchecked(DatabaseType::MySql, false)]
#[case::sqlite(DatabaseType::Sqlite, true)]
#[case::sqlite_unchecked(DatabaseType::Sqlite, false)]
fn only_runtime_bound_parameters(#[case] database_type: DatabaseType, #[case] checked: bool) {
    set_snapshot_suffix!(
        "{:?}{}",
        database_type,
        if checked { "" } else { "_unchecked" }
    );
    let input = quote::quote! {
        OutputType,
        r#"
            SELECT column
            FROM table
            WHERE created_at > {created_at}
        "#,
    };
    let output = crate::conditional_query_as(database_type, input, checked).unwrap();
    insta::assert_snapshot!(prettyprint(output));
}

#[rstest::rstest]
#[case::postgres(DatabaseType::PostgreSql, true)]
#[case::postgres_unchecked(DatabaseType::PostgreSql, false)]
#[case::mysql(DatabaseType::MySql, true)]
#[case::mysql_unchecked(DatabaseType::MySql, false)]
#[case::sqlite(DatabaseType::Sqlite, true)]
#[case::sqlite_unchecked(DatabaseType::Sqlite, false)]
fn only_compile_time_bound_parameters(#[case] database_type: DatabaseType, #[case] checked: bool) {
    set_snapshot_suffix!(
        "{:?}{}",
        database_type,
        if checked { "" } else { "_unchecked" }
    );
    let hash = proc_macro2::Punct::new('#', proc_macro2::Spacing::Alone);
    let input = quote::quote! {
        OutputType,
        r#"
            SELECT column
            FROM table
            WHERE value = {#value}
        "#,
        #hash value = match value {
            _ => "value",
        },
    };
    let output = crate::conditional_query_as(database_type, input, checked).unwrap();
    insta::assert_snapshot!(prettyprint(output));
}

#[rstest::rstest]
#[case::postgres(DatabaseType::PostgreSql, true)]
#[case::postgres_unchecked(DatabaseType::PostgreSql, false)]
#[case::mysql(DatabaseType::MySql, true)]
#[case::mysql_unchecked(DatabaseType::MySql, false)]
#[case::sqlite(DatabaseType::Sqlite, true)]
#[case::sqlite_unchecked(DatabaseType::Sqlite, false)]
fn both_parameter_kinds(#[case] database_type: DatabaseType, #[case] checked: bool) {
    set_snapshot_suffix!(
        "{:?}{}",
        database_type,
        if checked { "" } else { "_unchecked" }
    );
    let hash = proc_macro2::Punct::new('#', proc_macro2::Spacing::Alone);
    let input = quote::quote! {
        OutputType,
        r#"
            SELECT column
            FROM table
            WHERE
                created_at > {created_at}
                AND value = {#value}
        "#,
        #hash value = match value {
            _ => "value",
        },
    };
    let output = crate::conditional_query_as(database_type, input, checked).unwrap();
    insta::assert_snapshot!(prettyprint(output));
}
