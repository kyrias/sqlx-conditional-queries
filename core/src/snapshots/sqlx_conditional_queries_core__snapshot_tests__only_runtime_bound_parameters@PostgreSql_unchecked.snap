---
source: core/src/snapshot_tests.rs
expression: prettyprint(output)
---
fn dummy() {
    {
        enum ConditionalMap<'q, DB: ::sqlx::Database, A, F0> {
            Variant0(::sqlx::query::Map<'q, DB, F0, A>),
        }
        impl<'q, DB, A, O, F0> ConditionalMap<'q, DB, A, F0>
        where
            DB: ::sqlx::Database,
            A: 'q + ::sqlx::IntoArguments<'q, DB> + ::std::marker::Send,
            O: ::std::marker::Unpin + ::std::marker::Send,
            F0: ::std::ops::FnMut(DB::Row) -> ::sqlx::Result<O> + ::std::marker::Send,
        {
            /// See [`sqlx::query::Map::fetch`]
            pub fn fetch<'e, 'c: 'e, E>(
                self,
                executor: E,
            ) -> ::sqlx_conditional_queries::exports::BoxStream<'e, ::sqlx::Result<O>>
            where
                'q: 'e,
                E: 'e + ::sqlx::Executor<'c, Database = DB>,
                DB: 'e,
                O: 'e,
                F0: 'e,
            {
                match self {
                    Self::Variant0(map) => map.fetch(executor),
                }
            }
            /// See [`sqlx::query::Map::fetch_many`]
            #[deprecated = "Only the SQLite driver supports multiple statements in one prepared statement and that behavior is deprecated. Use `sqlx::raw_sql()` instead. See https://github.com/launchbadge/sqlx/issues/3108 for discussion."]
            pub fn fetch_many<'e, 'c: 'e, E>(
                mut self,
                executor: E,
            ) -> ::sqlx_conditional_queries::exports::BoxStream<
                'e,
                ::sqlx::Result<::sqlx::Either<DB::QueryResult, O>>,
            >
            where
                'q: 'e,
                E: 'e + ::sqlx::Executor<'c, Database = DB>,
                DB: 'e,
                O: 'e,
                F0: 'e,
            {
                match self {
                    Self::Variant0(map) => #[allow(deprecated)] map.fetch_many(executor),
                }
            }
            /// See [`sqlx::query::Map::fetch_all`]
            pub async fn fetch_all<'e, 'c: 'e, E>(
                self,
                executor: E,
            ) -> ::sqlx::Result<::std::vec::Vec<O>>
            where
                'q: 'e,
                E: 'e + ::sqlx::Executor<'c, Database = DB>,
                DB: 'e,
                O: 'e,
                F0: 'e,
            {
                match self {
                    Self::Variant0(map) => map.fetch_all(executor).await,
                }
            }
            /// See [`sqlx::query::Map::fetch_one`]
            pub async fn fetch_one<'e, 'c: 'e, E>(self, executor: E) -> ::sqlx::Result<O>
            where
                'q: 'e,
                E: 'e + ::sqlx::Executor<'c, Database = DB>,
                DB: 'e,
                O: 'e,
                F0: 'e,
            {
                match self {
                    Self::Variant0(map) => map.fetch_one(executor).await,
                }
            }
            /// See [`sqlx::query::Map::fetch_optional`]
            pub async fn fetch_optional<'e, 'c: 'e, E>(
                self,
                executor: E,
            ) -> ::sqlx::Result<::std::option::Option<O>>
            where
                'q: 'e,
                E: 'e + ::sqlx::Executor<'c, Database = DB>,
                DB: 'e,
                O: 'e,
                F0: 'e,
            {
                match self {
                    Self::Variant0(map) => map.fetch_optional(executor).await,
                }
            }
        }
        match ((),) {
            ((),) => {
                ConditionalMap::Variant0(
                    ::sqlx::query_as_unchecked!(
                        OutputType,
                        "\n            SELECT column\n            FROM table\n            WHERE created_at > "
                        + "$1" + "\n        ", created_at
                    ),
                )
            }
        }
    }
}
